### DISCUSSION TITLE
How to trigger a build using Github Actions from strapi webhook

### DISCUSSION THREAD

**Discussion by jgrau** [strapi/strapi#6576] [Legacy Topics]
Hi all

I am investigating whether to use Strapi and Gatsby for an upcoming project. I haven't used Strapi before but so far it looks like an exact match for what I'm looking for!

My plan is to use Github Actions with a pipeline that runs `gatsby build` and then builds an nginx docker container with the static files and pushed that to my image registry where another system takes over. I've successfully created a pipeline that can be triggered via the github api, however, the payload when posting to the github api has to look similar to

```
curl -X POST https://api.github.com/repos/:owner/:repo/dispatches \
-H 'Accept: application/vnd.github.everest-preview+json' \
-H 'Authorization: token $TOKEN' \
--data '{"event_type": "$CUSTOM_ACTION_NAME"}'
```

which I don't see is possible via the Strapi Webhooks functionality.

My question is:

Is there a way to modify the payload sent by the webhook? Or is there another way to accomplish the task? I was considering creating a route in strapi that the webhook would hit and then make that route hit the github api but as I'm new to Strapi I'm not sure whether that's a bad idea :)

**Comment by waynehaffenden** [comment]
I'm also trying to do exactly this, it seems that it's not possible. We might have to create a custom lifecycle method for every single collection type manually triggering the POST request to Github. I'm waiting on an answer too but this isn't looking hopeful. I'll update you if I make any progress

**Comment by chmielulu** [comment]
Hi! 

I've encountered a similar problem. Unfortunately, I didn't find it possible to add the body to the POST request, and the Github API requires this. So I decided to use [Buddy](https://buddy.works/). Free plan is very limited and wasn't enough for my requirements, but I used Buddy to trigger a request to Github API. Buddy is very easy to set up, because everything is clickable. To run Buddy's pipeline you don't need any information in the body of the request, so there is no problem with this in Strapi.

**So, my solution looks like:**
1. Create new pipeline in [Buddy](https://buddy.works/) with function "Send HTTP request", which will execute request to the Github API.
2. Copy webhook from [Buddy](https://buddy.works/) and add it to Strapi
3. Now, when you want to change something on a website in Strapi CMS, strapi will call a pipeline to Buddy, which in turn will call a pipeline in Github Actions.

**Comment by ggirodda** [comment]
I'm new to strapi , but I think I will face the same problem. I want to know if it's easier to create a custom route in strapi, triggered by the webhook, and do the request to github actions from the custom controller. Maybe it's better than using another tool ?

**Reply by ggirodda** [reply]
I tried it today, and everything works well. I created a strapi custom route, called by the webhook, and from there I call the github api using the `https://api.github.com/repos/:owner/:repo/dispatches` route.

**Reply by kennyvallejoc** [reply]
@ggirodda do you create a api route as  webhook, and inside that route you call the dispatch?

Its a shame that this technology dont allow u to do a lot of things :/

**Reply by ggirodda** [reply]
yes, exactly, and as I already have a service account authentication, using api key in a `X-API-Key` header, I created a user with permissions only on this route, and put the header in the webhook config. For now it's not so difficult to implement as a workaround, I wouldn't say that it's a shame, I think that for now they have others priorities

**Reply by kennyvallejoc** [reply]
thats what i though, thanks buddy.

**Reply by kwiat1990** [reply]
> I tried it today, and everything works well. I created a strapi custom route, called by the webhook, and from there I call the github api using the `https://api.github.com/repos/:owner/:repo/dispatches` route.

@ggirodda , could you elaborate more on how you managed to do this exactly? I tired to create a route in Strapi, which would then post to a Github action URL but it didn't work for me. I couldn't get the endpoint to work.

**Reply by ggirodda** [reply]
First you have to create a user that has the rights on your custom route. Than you have to create a service-account collection, with `apiKey` and `user` fields. the apiKey is just a string used for authentication (instead of the password), the user is a 1-1 or a n-1 relation with user (I used a 1-1). you also have to edit `extensions/users-permissions/config/policies/permissions.js` to allow auth by token, you can read [this post](https://strapi.io/documentation/developer-docs/latest/guides/api-token.html#add-token-validation-logic). Finally you put the url of your custom route in webhook config.
this is my `extensions/users-permissions/config/policies/permissions.js` file, I preferred to use x-api-key header instead of token query param.
```javascript
const _ = require('lodash');

module.exports = async (ctx, next) => {
  let role;

  if (ctx.state.user) {
    // request is already authenticated in a different way
    return next();
  }
  if (ctx.request && ctx.request.header && (ctx.request.header.authorization || ctx.request.header["x-api-key"])) {
    let id
    if (ctx.request.header["x-api-key"]) {
      const [serviceAccount] = await strapi.query('service-accounts').find({ apiKey: ctx.request.header["x-api-key"] });

      if (!serviceAccount) {
        throw new Error(`Invalid serviceAccount: This service account doesn't exist`);
      } else {
        if (serviceAccount.user && typeof serviceAccount.apiKey === 'string') {
          id = serviceAccount.user.id;
        }
      }
    } else if (ctx.request && ctx.request.header && ctx.request.header.authorization) {
      try {
        const user = await strapi.plugins['users-permissions'].services.jwt.getToken(ctx);
        id = user.id
      } catch (err) {
        return handleErrors(ctx, err, 'unauthorized');
      }
    }

    try {

      if (id === undefined) {
        throw new Error('Invalid token: Token did not contain required fields');
      }

      // fetch authenticated user
      ctx.state.user = await strapi.plugins[
        'users-permissions'
      ].services.user.fetchAuthenticatedUser(id);
    } catch (err) {
      return handleErrors(ctx, err, 'unauthorized');
    }

    if (!ctx.state.user) {
      return handleErrors(ctx, 'User Not Found', 'unauthorized');
    }

    role = ctx.state.user.role;

    if (role.type === 'root') {
      return await next();
    }

    const store = await strapi.store({
      environment: '',
      type: 'plugin',
      name: 'users-permissions',
    });

    if (
      _.get(await store.get({ key: 'advanced' }), 'email_confirmation') &&
      !ctx.state.user.confirmed
    ) {
      return handleErrors(ctx, 'Your account email is not confirmed.', 'unauthorized');
    }

    if (ctx.state.user.blocked) {
      return handleErrors(
        ctx,
        'Your account has been blocked by the administrator.',
        'unauthorized'
      );
    }
  }

  // Retrieve `public` role.
  if (!role) {
    role = await strapi.query('role', 'users-permissions').findOne({ type: 'public' }, []);
  }

  const route = ctx.request.route;
  const permission = await strapi.query('permission', 'users-permissions').findOne(
    {
      role: role.id,
      type: route.plugin || 'application',
      controller: route.controller,
      action: route.action,
      enabled: true,
    },
    []
  );

  if (!permission) {
    return handleErrors(ctx, undefined, 'forbidden');
  }

  // Execute the policies.
  if (permission.policy) {
    return await strapi.plugins['users-permissions'].config.policies[permission.policy](ctx, next);
  }

  // Execute the action.
  await next();
};

const handleErrors = (ctx, err = undefined, type) => {
  throw strapi.errors[type](err);
};
```

and this is my custom controller

```javascript
const axios = require('axios');

module.exports = {
  trigger: async ctx => {
    const githubUser = strapi.config.get('website-build.githubUser');
    const githubProject = strapi.config.get('website-build.githubProject');
    const githubBranch = strapi.config.get('website-build.githubBranch');
    const githubApiKey = strapi.config.get('website-build.githubApiKey');

    return axios.post(`https://api.github.com/repos/${githubUser}/${githubProject}/dispatches`,
      {
        event_type: "build-website",
        client_payload: { branch: githubBranch }
      },
      {
        headers: {
          Authorization: `token ${githubApiKey}`,
          Accept: "application/vnd.github.v3+json",
        }
      }
    )
  },
};
```
don't forget to add x-api-key header in header params of your webhook config, and put the value of the apiKey of the user that has the right to post on your custom route.

**Comment by kwiat1990** [comment]
@ggirodda thanks a lot for sharing your code. I will try your solution. I wonder what would be better - this custom made or something what involves 3rd party services like buddy. I hope Strapi will make it possible to add body to a webhook.

**Reply by ggirodda** [reply]
in my case I already used a service-account auth for other routes, it's only in a second step that I used it also to authenticate the call of the webhook. I didn't try yet, but maybe create a plugin it's the best solution

**Comment by kwiat1990** [comment]
@ggirodda  somehow I get `Error: The model service-accounts can't be found.` and thereby 500 error in frontend when I use the "trigger" option in Webhooks tab. The `service-account` collection does exist, so if there is any mistake it must be then in your code for `permission.js`, where `await strapi.query("service-accounts").find({ apiKey: ctx.request.header["x-api-key"] })` is used. 

I assume this should work on localhost as well.
